src/main/java/updater/DownloadCountingOutputStream.java

```diff
@@ -0,0 +1,35 @@
+package updater;
+
+import org.apache.commons.io.output.CountingOutputStream;
+
+import java.io.IOException;
+import java.io.OutputStream;
+import java.util.Optional;
+
+/**
+ * This class is the subject in observer pattern where the observer is a progress tracker.
+ */
+public class DownloadCountingOutputStream extends CountingOutputStream {
+    private final ProgressTracker progressTracker;
```

Comment:
Rather than take a concrete UI element to mutate, maybe this class could publish download progress in some other way (a callback, or an observable property, JavaFX-style), which another class could then make use of. Makes things a bit more decoupled.